adr/0003-partial-value-masking.md

```diff
@@ -0,0 +1,116 @@
+# 3. Partial value masking
+
+Date: 09-03-2024
+
+## Status
+
+Draft
+
+## Context
+
+Currently, the mask for numbers and string can be configured to be:
+
+1. A fixed-length string mask (e.g. `***`, `[REDACTED]`, etc.)
+2. A per-character mask for each character in the value (e.g. `#`, `*`, etc.) which results in a length-preserving mask
+3. Additionally, for numbers: type-preserving number mask (e.g. `0`, `8`, etc.)
+
+There have been multiple requests so far to support partial value masking on top of the aforementioned masking configurations. 
+
+Some examples of this are:
+
+- Partially masking an IBAN: `"NL91 FAKE 0417 1643 00"` as `"NL91 ************** 00"`
+- Partially masking an e-mail address: `"some-email-address@gmail.com"` as `"some-*******************l.com"`
+- Partially masking a phone number: `"+31 6 1234 7890"` as `"+31############890"`
```

Comment:
Also I wonder what the API should be for the phone one specifically. It can't be just `.leftVisibleCharacters(3)`, because 
1. There are country codes of different length(and a plus): `+380` (Ukraine) or `+1` (US)
2. I also wonder if we need a special handling when phone number doesn't start with `+` (local format), so the appropriate logic should be something like: 
    ```
    if starts with `+` - use the lookup table for country codes and keep it unmasked
    if starts with `0` - only keep `0` unmasked
    if starts with any digit other than `0` - mask everything (from the left)
    ```

Reply:
Phone number could also contain ()

Reply:
I think we can just keep it open honestly (using the Function<String, String> you described above).